### DISCUSSION TITLE
Scaling of the UI App

### DISCUSSION THREAD

**Discussion by vzdesic** [oqtane/oqtane.framework#911] [Q&A]
There is an **scaling issue with ASP.NET Core hosted Blazor**. It cannot be scaled because **SignalR is unable to communicate with the UI from the server**. That is why I want to use Blazor WebAssembly in my project. My UI must be scalable. 

**Is my UI scalable in case that I am using Oqtane (which also have server side) with Blazor WebAssembly?**

Thanks,
Vedran

**Comment by sbwalker** [comment]
@vzdesic I would encourage you to do some more research into Blazor, as the statement you made above is not accurate. There are many enterprise companies using Blazor Server in production in many different scenarios. The benchmarks that Microsoft shared for .NET Core 3.1 and Blazor Server are:

- A Standard_D1_V2 Azure instance (1 vCPU, 3.5 GB memory) can handle over 5,000 concurrent users without any degradation in latency.
- A Standard_D3_V2 Azure instance (4 vCPU, 14GB memory) handles well over 20,000 concurrent clients.

**Reply by vzdesic** [reply]
Your statement about performances is related to the single server as I understand. My challenge is that part of my application will handle very hard video streaming!

In Microsoft Docs ( https://docs.microsoft.com/en-us/aspnet/core/signalr/scale?view=aspnetcore-5.0 ) we can read following text/statement:
An app that uses SignalR needs to keep track of all its connections, which creates **problems for a server farm**. Add a server, and it gets new connections that the other servers don't know about. For example, SignalR on each server is unaware of the connections on the other servers. **When SignalR on one of the servers wants to send a message to all clients, the message only goes to the clients connected to that server.**

They gave two general solutions:
The options for solving this problem are the Azure SignalR Service and Redis backplane.

I just wanted to know if the same issue exists related to the Oqtane in any way? I am new to the Oqtane, but I see potential there, and I want to be sure what I am doing before my project starts.

Thanks,
Vedran

**Reply by sbwalker** [reply]
You could certainly use the Azure SignalR Service if you want to run Oqtane on Blazor Server. Or you could run on Blazor WebAssembly - in which case the client is downloaded to the browser and could connect to any server in the farm.

**Comment by sbwalker** [comment]
In regards to your other question, Oqtane can be hosted on Blazor Server or Blazor WebAssembly from a runtime perspective. There is a Runtime entry in the appsettings.json which controls this preference during app startup. If you specify "WebAssembly" then the UI will download/execute on WebAssembly in the browser. However like any other SPA application ( ie. client/server ) which is connected to a back-end database, Oqtane has a server-side API which runs on the server. If your intent is to create an application which is purely client-side and has no database or back-end services to consider ( ie. a game with no centralized high score tracking ), then Oqtane is not the right solution for you - you would be better off creating a pure client-side Blazor application from scratch.

**Reply by vzdesic** [reply]
My system will follow Microservices Architecture and Blazor WebAssembly will be technology used on the UI level. I will talk with my APIs from the UI. Part of the application will have to handle hard video streaming. Do you think I can benefit from Oqtane?

Also, important thing for me is to have Micro Frontends architecture on the UI. That is why I like Oqtane.

**Reply by PizzaConsole** [reply]
One of my big annoyances with oqtane in this regard is that you cannot decouple the Client side from the server side when you run the Webassembly version. Theoretically you should be able to host the client app as a static site and run your web api on a server. But because of how Oqtane is designed you cannot do that, it is dependent on the _host.cshtml file to be the index page

**Reply by mikecasas** [reply]
I am with you, I was trying to get to your point in my question #907, but you are more eloquent than me.

**Reply by sbwalker** [reply]
Early versions of Oqtane were configured to have an index.html in the Client project and an _host.cshtml in the Server project. In addition to these changes, you need to duplicate all Client static resources that are currently in the Server wwwroot to a wwwroot folder in the Client project. In order to support both Blazor hosting models with a single application, and avoid duplication of resources in both the Client and Server, we removed the wwwroot folder in the Client project and opted to bootstrapped WebAssembly using _host.cshtml. The other item to keep in mind is that some of the dynamic global JavaScript and CSS registration features are dependent on _host.cshtml as they rely on dynamic injection of elements to the initial page - which would be challenging to accomplish in a static index,html page. I am sure it would be possible to modify Oqtane to support a decoupled static site use case - with only minor degradation in features.

**Reply by PizzaConsole** [reply]
I understand you on this. IMHO Oqtane appears to be more designed with enterprise in mind. As a small freelancer I am having a hard time justifying using this framework over just creating a new blazor project from scratch. A big part of this for me is how you have to debug and build Oqtane. (Its very tedious) As nice as it is to have the awesome performance, "good enough" is what gets the job done and money in the door. Just some of my thoughts.

**Reply by sbwalker** [reply]
@PoisnFang that is a fair assessment. The benefits of modular frameworks are usually more applicable to larger applications. And it is certainly true that Oqtane has been designed with an enterprise mindset, and scalable architecture usually results in some added complexity - and this is where the balancing act comes into play. One of the goals of Oqtane is to make developers more productive. One aspect of this is by providing functional architecture and services out of the box which provide a substantial amount of the functionality required by most applications. The other aspect is the "inner loop" for developing and debugging custom extensions - and this is the area you are referring to which needs some optimization. I would agree - although part of the problem is that it has taken Microsoft quite some time to deliver a robust debugging experience for Blazor. Either way, if you are building small or simple custom applications for clients which do not require significant scale or future maintenance, then it may be a more attractive option to build them from scratch - although I would still argue it would be better in the long term to build them as modules.

**Comment by PizzaConsole** [comment]
> One of the goals of Oqtane is to make developers more productive

@sbwalker I understand the intend behind this statement. But given MY current experience in the framework. It does not come close to meeting this goal. I find something like Orchard Core, where they have turned it into a Nuget package, to be much more intuitive for starting a new site to give to a customer. (Though I find Orchard core to be way to monolithic for my taste) The front page of https://www.oqtane.org/ states "A Framework For Everyone" This is blatantly false, based on one of our previous discussions, where we specifically talked about how the barrier of entry for developing on Oqtane is Higher.

**I am not trying to be rude or diminish the brilliance of the framework, because it is quite incredible.** These are just my perspectives as a small freelance developer trying to figure out if Oqtane truly solves the problems I have while trying to develop in Blazor.

I realize that this is totally not on topic of this discussion post, but I don't really care to try and move it out.  (hmmm it didnt reply in the right part of the thread...)
